content/posts/2022-05-13-calculating-type-sets.markdown

```diff
@@ -0,0 +1,498 @@
+type NotZ interface{ Z() int }
+
+// (¬X ∨ Z) ∧ (Y ∨ Z) ∧ (X ∨ ¬Y)
+type S interface {
+    NotX | Z
+    Y | Z
+    X | NotY
+}
+```
+
+There is one detail which becomes important here. There are types which
+*neither* implement `X` *nor* do they implement `NotX` - if they don't have an
+`X` method at all, for example. If we want to really prove that calculating
+these new type sets is NP complete, we need to fix that. We want that a type
+set is empty *exactly* if its formula is not satisfiable. Boolean variables are
+always either `true` or `false`, though.
+
+“Luckily”, our new operator gives us a way to fix that:
+
+```go
+type TertiumNonDatur interface {
+    X | NotX
+    Y | NotY
+    Z | NotZ
+}
+
+// (¬X ∨ Z) ∧ (Y ∨ Z) ∧ (X ∨ ¬Y)
+type S interface {
+    TertiumNonDatur
+
+    NotX | Z
+    Y | Z
+    X | NotY
+}
+```
+
+Now, any type we consider necessarily has either an `X()` or an `X() int`
+method, so either implements `X`, or implements `NotX` (and never both).
+
+So, checking if these general type sets are empty is NP complete.
+
+Even worse, as the definition of what operations you can do on a variable of
+type parameter type, [we can show that *even type-checking a generic function*
+is NP-complete](https://github.com/golang/go/issues/45346#issuecomment-822330394)
+with this language.
+
+## Why do we care?
+
+A fair question is “why do we even care? Surely these cases are super exotic
+and in any real program, checking this is trivial”.
+
+That's true, but there are still reasons to care:
+
+- We still have to at least *talk about* type sets, because they also define
+  the set of valid operations on a type parameter constrained by it. We still
+  have to be sure that we can do all the necessary computations in the type
+  checker.
+- Go has the goal of having a fast compiler. And importantly, one which is
+  guaranteed to be fast *for any program*. If I give you a Go program, you can
+  be reasonably sure that it compiles quickly, in a time frame predictable by
+  the size of the input.
+
+  If I *can* craft a program which compiles slowly, this is no longer true.
+
+  This is especially important for environments like the Go playground, which
+  regularly compiles untrusted code.
+- NP complete problems are notoriously hard to debug, if they fail.
+
+  If you use Linux, you might have occasionally run into a problem where you
+  accidentally tried installing conflicting versions of some package. And if
+  so, you might have noticed that your computer first chugged along for a while
+  and then gave you an unhelpful error message about the conflict. And maybe
+  you had trouble figuring out which packages declared the conflicting
+  dependencies.
+
+  This is something vrey typical for NP complete problems. As an exact solution
+  is often too hard to compute, they rely on heuristics and randomization and
+  it's hard to work backwards from a failure.
+- It would require to somehow specify how a compiler should implement this
```

Comment:
This doesn't read great. What exactly would require the implementation specification? In general, the spec doesn't mention implementation algorithms, right?

Reply:
Rewrote this to make it clearer. In general, the spec does not *need* to mention an implementation algorithm, because there are correct and efficient choices. If there is no correct and efficient choice, we get the dichotomy "we have to prescribe a heuristic" or "programs might stop compiling when the heuristic changes or differs between compilers".